#162948837  Add descriptive Registration validation errors

Georgeygigz opened the pull request:
**What does this PR do?**

Add descriptive Registration validation errors

**Description of tasks to be completed?**

- Add descriptive Registration validation errors.
- Add test for validators

**How should this be manually tested?**
 - Start the server and navigate to this url `/api/users/`
- Use the following details to signup
```
{
  "user":{
    "username": "Jacob",
    "email": "jake@jake.jake",
    "password": "jakejake"
  }
}
```

**How to run tests**
- `python3 manage.py test`

**Relevant Pivotal Tracker Stories**

[#162948837](https://www.pivotaltracker.com/story/show/162948837)

**Screenshots**
![image](https://user-images.githubusercontent.com/24627269/50985798-3e49d280-1516-11e9-8d78-d5bcade2069f.png)

![image](https://user-images.githubusercontent.com/24627269/50985843-59b4dd80-1516-11e9-9dc3-936fe797d289.png)

![image](https://user-images.githubusercontent.com/24627269/50985861-63d6dc00-1516-11e9-8cdd-da241bbdac49.png)

Georgeygigz committed:
Feature(Validation errors):Registration validation errors
- Add descriptive validation signup errors
- Add tests for the validations

[Starts #162948837]

MaggieChege commented:
You've written tests. Please add how you can manually test them

kwameasiago reviewed (approved):
Looks Good maybe remember to add a link to the PT stories

d-kahara reviewed (approved):
Looks good to me.

KMaina reviewed (approved):
Looks good to me

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -1,4 +1,5 @@
 from django.contrib.auth import authenticate
+import re
```
Import should be in this order - python's packages first, followed by the framework's packages, then third-party packages and then your own packages/modules

Georgeygigz replied:
Okay, let me fix that.

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -8,13 +9,40 @@
+        error_messages={
+            'required': 'Password field required',
+            'min_length': 'Ensure Password field has at least 8 characters',
+            'invalid': 'password should contain a lowercase, uppercase numeric'
```
Start error message with capital letter just like the others.

Georgeygigz replied:
Okay

ignatiusukwuoma commented on authors/apps/authentication/tests/test_user_login.py:
```diff
@@ -11,14 +11,14 @@ def setUp(self):
         self.signup_data = {
             "user": {
-                "username": "gigz",
+                "username": "MaryGigz",
```
Why was the username for this test changed?

Georgeygigz replied:
Because I validated the username to have at least 6 characters.

ignatiusukwuoma commented on authors/apps/authentication/tests/test_user_registration.py:
```diff
@@ -137,3 +166,49 @@ def test_register_duplicate_username(self):
                                                          " username already"
                                                          " exists."]}})
+
+    def test_username_special_characters_only(self):
```
test names should say what the test is doing, example: `test_registration_rejects_username_with_special_characters_only`

Georgeygigz replied:
Okay noted.

ignatiusukwuoma commented on authors/apps/authentication/tests/test_user_registration.py:
```diff
@@ -137,3 +166,49 @@ def test_register_duplicate_username(self):
+                                    format="json")
+        self.assertEqual(json.loads(response.content), {'errors': {'password': 
+                         ['Ensure Password field has at least 8 characters']}})
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
No tests for invalid email

Georgeygigz replied:
Okay I will add them

Georgeygigz committed:
Feature(Adding more):Add more tests
- Test for invalid email
- Rewrting the order of imports
- Renaming the error message  to start with caps

[Delivers #162948837]

ignatiusukwuoma commented on authors/apps/authentication/tests/test_user_registration.py:
```diff
@@ -29,58 +29,92 @@ def setUp(self):
+                                                    "username": "MaryWnaja",
+                                                    "email": "jake1jake.jake",
+                                                    "password": "@Maty112Mna"
+                                                }} 
```
Leave a blank line before the test method

Georgeygigz replied:
Okay I am working on them

ignatiusukwuoma commented on authors/apps/authentication/tests/test_user_registration.py:
```diff
@@ -29,58 +29,92 @@ def setUp(self):
+                                                "username": "#@##@#@##@#",
+                                                "email": "jake@jake.jake",
+                                                "password": "g@AHJA473"
+                                                }}
```
The indentation of the user objects seems off... Can you just give the objects just four spaces? like Maggie's

Georgeygigz replied:
I am working on this too
